File: backend_python/app/routers/instruments.py
from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.orm import Session

from app.dependencies import get_current_user, get_db
from app.models.instrument import InvestmentType
from app.models.user import User
from app.schemas.instrument import InstrumentCreate, InstrumentPage, InstrumentRead, InstrumentUpdate, UserInstrumentRead
from app.services.instrument_service import (
    create_instrument,
    get_instrument,
    list_instrument_types,
    list_instruments_paged,
    list_tracked_instruments,
    list_user_instruments,
    track_instrument,
    untrack_instrument,
    update_instrument,
)

router = APIRouter(prefix="/api/v1/instruments", tags=["instruments"])


@router.get("", response_model=InstrumentPage)
def list_all(
    type: InvestmentType | None = Query(None),
    search: str | None = Query(None),
    cursor: int | None = Query(None, ge=1),
    limit: int = Query(50, ge=1, le=200),
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    return list_instruments_paged(db, type, search, cursor, limit)


@router.post("", response_model=InstrumentRead, status_code=status.HTTP_201_CREATED)
def create(
    body: InstrumentCreate,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    return create_instrument(db, body)


@router.get("/types", response_model=list[str])
def get_types(
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    return list_instrument_types(db)


@router.get("/tracked", response_model=list[InstrumentRead])
def list_tracked(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return list_tracked_instruments(db, current_user.id)


@router.get("/user-instruments", response_model=list[UserInstrumentRead])
def list_user_instruments_route(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return list_user_instruments(db, current_user.id)


@router.get("/{instrument_id}", response_model=InstrumentRead)
def get_one(
    instrument_id: int,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    return get_instrument(db, instrument_id)


@router.put("/{instrument_id}", response_model=InstrumentRead)
def update(
    instrument_id: int,
    body: InstrumentUpdate,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    return update_instrument(db, instrument_id, body)


@router.post("/{instrument_id}/track", status_code=status.HTTP_204_NO_CONTENT)
def track(
    instrument_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    track_instrument(db, instrument_id, current_user.id)


@router.delete("/{instrument_id}/track", status_code=status.HTTP_204_NO_CONTENT)
def untrack(
    instrument_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    untrack_instrument(db, instrument_id, current_user.id)

File: frontend/src/api/instruments.ts
import type {
  ApiResponse,
  Instrument,
  InstrumentCreate,
  InstrumentPage,
  InstrumentPositionSummary,
  InstrumentPricePoint,
  InvestmentType,
  LotRead,
  Transaction,
  UserInstrument,
} from '@/types'
import client from './client'

export async function listInstruments(
  params: { type?: InvestmentType; search?: string; cursor?: number; limit?: number } = {}
): Promise<InstrumentPage> {
  const res = await client.get<ApiResponse<Instrument[]>>('/instruments', { params })
  return {
    items:       res.data.data,
    next_cursor: (res.data.meta_data.next_cursor as number | null) ?? null,
    has_more:    (res.data.meta_data.has_more as boolean) ?? false,
  }
}

export async function listInstrumentTypes() {
  const res = await client.get<ApiResponse<string[]>>('/instruments/types')
  return res.data.data as InvestmentType[]
}

export async function listTrackedInstruments() {
  const res = await client.get<ApiResponse<Instrument[]>>('/instruments/tracked')
  return res.data.data
}

export async function createInstrument(data: InstrumentCreate) {
  const res = await client.post<ApiResponse<Instrument>>('/instruments', data)
  return res.data.data
}

export async function trackInstrument(id: number) {
  await client.post(`/instruments/${id}/track`)
}

export async function untrackInstrument(id: number) {
  await client.delete(`/instruments/${id}/track`)
}

export async function listUserInstruments() {
  const res = await client.get<ApiResponse<UserInstrument[]>>('/instruments/user-instruments')
  return res.data.data
}

// ── Profile endpoints ─────────────────────────────────────────────────────

export async function getInstrumentPosition(id: number) {
  const res = await client.get<ApiResponse<InstrumentPositionSummary>>(`/instruments/${id}/position`)
  return res.data.data
}

export async function getInstrumentLots(id: number) {
  const res = await client.get<ApiResponse<LotRead[]>>(`/instruments/${id}/lots`)
  return res.data.data
}

export async function getInstrumentTransactions(id: number, limit = 50) {
  const res = await client.get<ApiResponse<Transaction[]>>(
    `/instruments/${id}/transactions`,
    { params: { limit } },
  )
  return res.data.data
}

export async function getInstrumentPriceHistory(id: number, days = 90) {
  const res = await client.get<ApiResponse<InstrumentPricePoint[]>>(
    `/instruments/${id}/price-history`,
    { params: { days } },
  )
  return res.data.data
}
